rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp

```diff
@@ -0,0 +1,210 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <chrono>
+#include <cstdio>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rosbag2_compression/sequential_compression_reader.hpp"
+#include "rosbag2_compression/sequential_compression_writer.hpp"
+#include "rosbag2_cpp/converter_options.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/storage_options.hpp"
+#include "rosbag2_cpp/writer.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+#include "rosbag2_storage/topic_metadata.hpp"
+
+namespace rosbag2_py {
+
+class Reader {
+public:
+  Reader(const std::string &reader_class) {
+    if (!reader_class.compare("SequentialReader")) {
+      auto sequential_reader =
+          std::make_unique<rosbag2_cpp::readers::SequentialReader>();
+      reader_ =
+          std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else if (!reader_class.compare("SequentialCompressionReader")) {
+      auto sequential_reader =
+          std::make_unique<rosbag2_compression::SequentialCompressionReader>();
+      reader_ =
+          std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else {
+      throw std::runtime_error{"Reader class type " + reader_class +
+                               " not supported."};
+    }
+  }
+
+  void open(rosbag2_cpp::StorageOptions &storage_options,
+            rosbag2_cpp::ConverterOptions &converter_options) {
+    reader_->open(storage_options, converter_options);
+  }
+
+  bool has_next() { return reader_->has_next(); }
+
+  /// Return a tuple containing the topic name, the serialized ROS message, and
+  /// the timestamp.
+  pybind11::tuple read_next() {
+    const auto next = reader_->read_next();
+    rcutils_uint8_array_t rcutils_data = *next->serialized_data.get();
+    std::string serialized_data(
+        rcutils_data.buffer, rcutils_data.buffer + rcutils_data.buffer_length);
+    return pybind11::make_tuple(
+        next->topic_name, pybind11::bytes(serialized_data), next->time_stamp);
+  }
+
+  /// Return a mapping from topic name to topic type.
+  std::vector<rosbag2_storage::TopicMetadata> get_all_topics_and_types() {
+    return reader_->get_all_topics_and_types();
+  }
+
+private:
+  std::unique_ptr<rosbag2_cpp::Reader> reader_;
+};
+
+class SerializedBagMessage : public rosbag2_storage::SerializedBagMessage {
```

Comment:
Does the `SerializedBagMessage` struct need a wrapper class, or can it simply be wrapped with `def_readwrite()` like `TopicMetadata`? Then you would be able to use the `rosbag2_storage::SerializedBagMessage` directly. I don't know if that would solve the seg fault, but it wouldn't require the `memcpy` here which could lead to something like different pointer addresses. See the other comment for how I think it can be done.